### PR TITLE
add ECDSA signing and verification on the secp256k1 curve

### PR DESCRIPTION
**Pull request by polubelova**
This PR adds support for ECDSA signing and verification on the secp256k1 elliptic curve.

Future work:
- add GLV to efficiently compute the EC scalar multiplication
- prove lemmas needed for `Lib.Exponentiation` (e.g., `(points on the curve, aff_point_add)` forms a commutative monoid)

### CODE REVIEW

File: specs/Spec.K256.fst [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+  ecdsa_verify_hashed_msg msgHash public_key signature
+
+
+///  Parsing and Serializing public keys
+
+// raw          = [ x; y ], 64 bytes
+// uncompressed = [ 0x04; x; y ], 65 bytes
+// compressed   = [ 0x02 for even `y` and 0x03 for odd `y`; x ], 33 bytes
+
+let pk_uncompressed_to_raw (pk:lbytes 65) : option (lbytes 64) =
```
[the remark this answers is not in the document]
Comment (polubelova) [reply specs/Spec.K256.fst:165]: the secp256k1 has another curve equation so the algorithm of computing `y` from `x` is different. 
as for `uncompressed` public keys, we just need to copy the last 64 bytes, so it doesn't make much sense to share it with p256.

**Comment by polubelova** [comment]
@msprotz thanks for your comments!

That's right, this PR also introduces some changes to exponentiation. 

There are three files: `lib/Lib.Exponentiation` (high-level specification), `specs/Spec.Exponentiation` (low-level specification) and `code/bignum/Hacl.Impl.Exponentiation` (low* implementation). 
Until now, `Hacl.Impl.Exponentiaion` was functionally correct with respect to `Lib.Exponentiation` without any link to `Spec.Exponentiation`. It worked fine for EC scalar multiplication of Ed25519 and modular exponentiation. 

The [bitcoin-core/secp256k1](https://github.com/bitcoin-core/secp256k1/blob/master/include/secp256k1.h) implementation has a [nice optimization](https://github.com/bitcoin-core/secp256k1/blob/912b7ccc4473b5c969b01d027b8d5dc515435eb5/src/ecdsa_impl.h#L244) that avoids computing the modular multiplicative inverse (which is an expensive operation) for `ecdsa-verify` when `the order of the curve < prime` and `prime < 2 * the order of the curve`. To [prove](https://github.com/project-everest/hacl-star/blob/f384d5c8b2f438cb50f63226b83a61aedd15d06d/specs/lemmas/Spec.K256.Lemmas.fsti#L37) the correctness of such optimization we need to continue working with projective (= low-level specification) coordinates rather than with affine (= high-level specification) ones. That's why `Hacl.Impl.Exponentiation` is now proven to be functionally correct against `Spec.Exponentiation`. 

I hope this explains the changes in the library.

**Comment by polubelova** [comment]
@karthikbhargavan I don't see atm how one can easily write a generic specification that works for both curves.

The main difference is that our specification of secp256k1 uses projective coordinates (needed for the complete point addition and doubling formulas given in [this paper](https://eprint.iacr.org/2015/1060.pdf)), while p256 uses jacobian ones.

As for `ecdsa-verify` and `ecdsa-sign`, they basically load inputs into a field or scalar representation, compute EC scalar multiplication and serialize/check the result. As I've already said, there is a [nice optimization](https://github.com/bitcoin-core/secp256k1/blob/912b7ccc4473b5c969b01d027b8d5dc515435eb5/src/ecdsa_impl.h#L244) for `ecdsa-verify` that avoids computing a modular multiplicative inverse (needed for converting a point from the projective to affine coordinates), so the EC scalar multiplication API no longer works with affine coordinates.

@Frosne, it seems that the above optimization also works for p256 as `order_p256 < prime_p256` and `prime_p256 < 2 * order_p256`. If so, it would be great to implement it for `p256-ecdsa-verify` as well, as it saves two modular multiplicative inverse operations from [`_norm`](https://github.com/project-everest/hacl-star/blob/0bd6f56e0a4b98f930ceaed1959d859428b63365/specs/ecdsap256/Spec.P256.fst#L84).

**Review by msprotz — APPROVED** [review]
This is good for me, thanks for adding the explanations!

**Comment by Frosne** [comment]
I will put it to my to-do list (:

**Pull request merged**